drivers/i2c/i2c_nrfx_twim.c

```diff
@@ -336,6 +336,9 @@ static int i2c_nrfx_twim_init(const struct device *dev)
 
 	dev_config->irq_connect();
 
+	k_sem_init(&dev_data->transfer_sync, 1, 1);
```

Comment:
Why move to dynamic initialization of the semaphores?

Reply:
For the reason described in the commit message. Any RAM variables that have default values need to be stored in their entirety in the `.data` section. For example this pattern would use an extra 1kB of ROM.
```
struct dev_data {
    int counter;
    uint8_t buffer[1024];
}

struct dev_data my_dev_data = {.counter = 2};
```
I'm not claiming the I2C usage is that bad, but `k_sem_init` is basically guaranteed to already exist in the image, so this is just free ROM savings. By my count the absolute minimum size for the data structure is 36 bytes, increasing further if any of `CONFIG_POLL`, `CONFIG_TRACING` or `CONFIG_OBJ_CORE_SEM` are enabled.